File: ransomeye_intelligence/security/trust_chain.py
# Path and File Name : /home/ransomeye/rebuild/ransomeye_intelligence/security/trust_chain.py
# Author: nXxBku0CKFAJCBN3X1g3bQk7OxYQylg8CMw1iGsq7gU
# Details of functionality of this file: Validates trust chain for intelligence artifacts - verifies certificate chain

"""
Trust Chain Validation: Validates trust chain for intelligence artifacts.
Verifies certificate chain against Root CA.
"""

from pathlib import Path
from typing import Dict, List, Tuple
from ransomeye_trust.verify_tool import VerifyTool


class IntelligenceTrustChain:
    """Validates trust chain for intelligence artifacts."""
    
    def __init__(self, trust_dir: str = "/home/ransomeye/rebuild/ransomeye_trust"):
        self.trust_dir = Path(trust_dir)
        self.verify_tool = VerifyTool(str(self.trust_dir))
    
    def validate_trust_chain(self, artifact_path: Path, domain: str = "artifacts") -> Tuple[bool, List[str]]:
        """
        Validate trust chain for artifact.
        
        Args:
            artifact_path: Path to artifact
            domain: Trust domain
        
        Returns:
            Tuple of (is_valid: bool, errors: List[str])
        """
        errors = []
        
        # Verify certificate chain
        try:
            certificate = self.verify_tool.load_certificate(domain)
            if not self.verify_tool.verify_certificate_chain(certificate):
                errors.append(f"Certificate chain validation failed for {artifact_path.name}")
        except Exception as e:
            errors.append(f"Error validating trust chain: {e}")
        
        return len(errors) == 0, errors


def main():
    """CLI entry point for trust chain validator."""
    trust_chain = IntelligenceTrustChain()
    
    # Example validation
    artifact_path = Path("/home/ransomeye/rebuild/ransomeye_intelligence/baseline_pack/models/model_manifest.json")
    is_valid, errors = trust_chain.validate_trust_chain(artifact_path)
    
    if is_valid:
        print("✓ Trust chain validation passed")
    else:
        print("✗ Trust chain validation failed:")
        for error in errors:
            print(f"  {error}")


if __name__ == '__main__':
    main()

File: ransomeye_ingestion/src/security/trust_chain.rs
// Path and File Name : /home/ransomeye/rebuild/ransomeye_ingestion/src/security/trust_chain.rs
// Author: nXxBku0CKFAJCBN3X1g3bQk7OxYQylg8CMw1iGsq7gU
// Details of functionality of this file: Trust chain validation - validates certificate chains against root CA with real cryptography

/*
 * Trust Chain Validation
 * 
 * Validates certificate chains against root CA.
 * Ensures trust chain integrity using real cryptographic verification.
 * Fail-closed: rejects on any chain validation failure.
 */

use std::sync::Arc;
use x509_parser::prelude::*;
use ring::signature::{UnparsedPublicKey, RSA_PSS_SHA256};
use sha2::{Sha256, Digest};
use tracing::{error, debug, warn};

use crate::security::errors::IdentityError;
use crate::security::trust_store::TrustStore;

pub struct TrustChainValidator {
    trust_store: Arc<TrustStore>,
}

impl TrustChainValidator {
    pub fn new(trust_store: Arc<TrustStore>) -> Result<Self, IdentityError> {
        Ok(Self {
            trust_store,
        })
    }
    
    /// Validate certificate chain against root CA
    /// Returns Ok(()) on success, IdentityError on failure
    pub async fn validate_chain(&self, certificate_der: &[u8]) -> Result<(), IdentityError> {
        debug!("Validating certificate chain");
        
        // Parse certificate
        let (_, cert) = X509Certificate::from_der(certificate_der)
            .map_err(|e| IdentityError::CertificateParseError(
                format!("Failed to parse certificate: {}", e)
            ))?;
        
        // Get root CA
        let root_ca = self.trust_store.get_root_ca()?;
        
        // Verify certificate is signed by root CA
        self.verify_signature(&cert, &root_ca)?;
        
        // Verify issuer matches root CA subject
        let cert_issuer = cert.issuer().to_string();
        let root_subject = root_ca.subject().to_string();
        
        if cert_issuer != root_subject {
            return Err(IdentityError::ChainValidationFailed(
                format!("Certificate issuer '{}' does not match root CA subject '{}'", 
                    cert_issuer, root_subject)
            ));
        }
        
        debug!("Certificate chain validated successfully");
        Ok(())
    }
    
    fn verify_signature(&self, cert: &X509Certificate, issuer_cert: &X509Certificate) -> Result<(), IdentityError> {
        // Get issuer's public key
        let issuer_public_key = issuer_cert.public_key();
        let issuer_public_key_bytes = issuer_public_key.raw;
        
        // Get certificate's signature algorithm
        let sig_alg = cert.signature_algorithm();
        
        // Verify signature algorithm is RSA-PSS-SHA256 or RSA-PKCS1-SHA256
        let algorithm = match sig_alg.algorithm().to_id_string().as_str() {
            "1.2.840.113549.1.1.10" => &RSA_PSS_SHA256, // RSA-PSS
            "1.2.840.113549.1.1.11" => &RSA_PSS_SHA256, // RSA-PSS with SHA-256 (approximate)
            _ => {
                warn!("Unsupported signature algorithm: {}", sig_alg.algorithm().to_id_string());
                // Try RSA-PSS-SHA256 as default
                &RSA_PSS_SHA256
            }
        };
        
        // Get certificate's TBS (To Be Signed) certificate bytes
        // This is the certificate data that was signed
        let tbs_certificate = cert.tbs_certificate.as_ref();
        
        // Compute hash of TBS certificate
        let mut hasher = Sha256::new();
        hasher.update(tbs_certificate);
        let hash = hasher.finalize();
        
        // Get certificate signature
        let signature = cert.signature_value();
        
        // Verify signature
        let unparsed_key = UnparsedPublicKey::new(algorithm, issuer_public_key_bytes);
        
        unparsed_key.verify(&hash, signature.as_ref())
            .map_err(|e| IdentityError::ChainValidationFailed(
                format!("Certificate signature verification failed: {}", e)
            ))?;
        
        Ok(())
    }
}
